bottomless/src/read.rs

```diff
@@ -55,3 +58,173 @@ impl BatchReader {
+                )
+                .send()
+                .await?;
+
+            Ok(())
+        }
+        Err(err) => {
+            client
+                .abort_multipart_upload()
+                .upload_id(upload_id)
+                .bucket(bucket)
+                .key(key)
+                .send()
+                .await?;
+
+            Err(err)
+        }
+    }
+}
+
+async fn upload_s3_parts(
+    client: &aws_sdk_s3::Client,
+    upload_id: &str,
+    key: &str,
+    bucket: &str,
+    mut reader: impl AsyncRead + Unpin,
+) -> Result<Vec<CompletedPart>> {
+    let chunk_sizes = &[
+        5 * 1024 * 1024,
+        10 * 1024 * 1024,
+        25 * 1024 * 1024,
+        50 * 1024 * 1024,
+        100 * 1024 * 1024,
+    ];
+
+    const LAST_PART: i32 = 10_000;
+    let mut parts = Vec::new();
+    let mut has_reached_eof = false;
+
+    // S3 allows a maximum of 10_000 parts and each part can size from 5 MiB to
+    // 5 GiB, except for the last one that has no limits.
+    //
+    // See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
+    for part in 0..LAST_PART - 1 {
+        // Progressively increase the chunk size every 16 chunks up to the last
+        // chunk_size. This allows smaller allocations for small databases.
+        //
+        // Here's a table of how much data we can chunk:
+        // ┌────────────┬──────────────────┬───────────────────────┬──────────────────┐
+        // │ Chunk size │ Number of chunks │ Amount for chunk size │ Cumulative total │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 5 MiB      │ 16               │ 80 MiB                │ 80 MiB           │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 10 MiB     │ 16               │ 160 MiB               │ 240 MiB          │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 25 MiB     │ 16               │ 400 MiB               │ 640 MiB          │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 50 MiB     │ 16               │ 800 MiB               │ 1.406 GiB        │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 100 MiB    │ 9935             │ 970.215 GiB           │ 971.621 GiB      │
+        // └────────────┴──────────────────┴───────────────────────┴──────────────────┘
+        //
+        // We can send up to 971 GiB in chunks, which is more than enough for the
+        // majority of use cases.
+        //
+        // The last chunk is reserved for the remaining of the `gzip_reader`
+        let chunk_size = chunk_sizes[((part / 16) as usize).min(chunk_sizes.len() - 1)];
+
+        let mut buffer = bytes::BytesMut::with_capacity(chunk_size);
+        loop {
+            let bytes_written = reader.read_buf(&mut buffer).await?;
```

Comment:
Ideally we should probably use `AsyncBufRead` trait on the reader and use its internal buffer instead to avoiding copying between multiple intermediate buffers - I know however that `async_compression` crate may be inflexible in this regard.

Reply:
I tried to do that but unfortunally the S3 SDK is not very flexible. `ByteStream` can wrap a [`aws_smithy_http::body::SdkBody`](https://docs.rs/aws-smithy-http/latest/aws_smithy_http/body/struct.SdkBody.html) which requires that you return a `bytes::Byte` at some point and that would require copying the buffer into it anyway.

The only way I can think of to make this zero-copy is to use the S3 REST API.